neuron.py

```diff
@@ -1,41 +1,45 @@
 import numpy as np
 from numpy import dot, ndarray
 
+from activation_functions import ActivationFunctionType
+from error_functions import ErrorFunctionType
+
 
 class Neuron:
-    def __init__(self, activation_function: callable) -> None:
+    def __init__(self, activation_function: ActivationFunctionType, error_function: ErrorFunctionType) -> None:
         self.weights = []
         self.delta = 0
         self.neuron_index_in_layer = 0
         self.__activation_function = activation_function
+        self.__error_function = error_function
         self.__input = None
-        self.__output = 0
+        self.__output = 0.0
 
     def set_neuron_number_in_layer(self, index: int) -> None:
         self.neuron_index_in_layer = index
 
     def copy_neuron(self) -> 'Neuron':
-        return Neuron(self.__activation_function)
+        return Neuron(self.__activation_function, self.__error_function)
 
-    def set_weights(self, weights) -> None:
+    def set_weights(self, weights: list[float]) -> None:
         self.weights = weights
 
     def get_weights(self) -> list[float]:
         return self.weights
 
     def calculate_output(self, input_value: ndarray) -> float:
         self.__input = input_value
-        self.__output = self.__activation_function(dot(self.weights, input_value))
+        self.__output: float = self.__activation_function(dot(self.weights, input_value))
         return self.__output
 
     # Based on: https://machinelearningmastery.com/implement-backpropagation-algorithm-scratch-python/
     def calculate_error(self, expected_output: list[float], next_layer) -> float:
         error = 0
         if next_layer is None:
-            error = self.__output - expected_output[self.neuron_index_in_layer]
+            error = -self.__error_function(self.__output, expected_output[self.neuron_index_in_layer], True)
```

Comment:
Where does this minus before calculating error function comes from? This value is later also multiplied by `-1` when updating weights. If this is because results get better then I have a sense maybe there is an issue with error function derivative, but it is just a guess

Reply:
Minus is necessary but I've moved it to functions definitions